### PR TITLE
TS-4627: support TSRemapOSResponse in ts_lua

### PR DESCRIPTION
**Pull request by shukitchan**
@jpeach pls take a look. Thanks.

I want to add enhancement to ts_lua so I can write script to take advantage of the TSRemapOSResponse() related functionality.

Basically it would allow me to do something when origin server fails to connect (such as DNS error, timeout, etc)

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/398/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/294/ for details.

File: plugins/experimental/ts_lua/ts_lua_http.c [hunk 1]
```diff
@@ -29,9 +29,33 @@ typedef enum {
...
                                              "TS_LUA_CACHE_LOOKUP_HIT_FRESH", "TS_LUA_CACHE_LOOKUP_SKIPPED"};
 
+char *ts_lua_server_state_string[] = {"TS_LUA_SRVSTATE_STATE_UNDEFINED", "TS_LUA_SRVSTATE_ACTIVE_TIMEOUT"
+                                                                         "TS_LUA_SRVSTATE_BAD_INCOMING_RESPONSE",
+                                      "TS_LUA_SRVSTATE_CONNECTION_ALIVE"
+                                      "TS_LUA_SRVSTATE_CONNECTION_CLOSED",
+                                      "TS_LUA_SRVSTATE_CONNECTION_ERROR", "TS_LUA_SRVSTATE_INACTIVE_TIMEOUT",
+                                      "TS_LUA_SRVSTATE_OPEN_RAW_ERROR", "TS_LUA_SRVSTATE_PARSE_ERROR",
+                                      "TS_LUA_SRVSTATE_TRANSACTION_COMPLETE", "TS_LUA_SRVSTATE_CONGEST_CONTROL_CONGESTED_ON_F",
+                                      "TS_LUA_SRVSTATE_CONGEST_CONTROL_CONGESTED_ON_M"};
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua_http.c]: Add a ',' before the closing brace to encourage ``clang-format` to format this right.

File: plugins/experimental/ts_lua/ts_lua_http.c [hunk 2]
```diff
@@ -29,9 +29,33 @@ typedef enum {
...
+  TS_LUA_SRVSTATE_PARSE_ERROR,
+  TS_LUA_SRVSTATE_TRANSACTION_COMPLETE,
+  TS_LUA_SRVSTATE_CONGEST_CONTROL_CONGESTED_ON_F,
+  TS_LUA_SRVSTATE_CONGEST_CONTROL_CONGESTED_ON_M
+} TSLuaServerState;
+
 char *ts_lua_cache_lookup_result_string[] = {"TS_LUA_CACHE_LOOKUP_MISS", "TS_LUA_CACHE_LOOKUP_HIT_STALE",
                                              "TS_LUA_CACHE_LOOKUP_HIT_FRESH", "TS_LUA_CACHE_LOOKUP_SKIPPED"};
 
+char *ts_lua_server_state_string[] = {"TS_LUA_SRVSTATE_STATE_UNDEFINED", "TS_LUA_SRVSTATE_ACTIVE_TIMEOUT"
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua_http.c]: ``const char *``

File: plugins/experimental/ts_lua/ts_lua.c [hunk 3]
```diff
@@ -116,6 +116,69 @@ TSRemapDeleteInstance(void *ih)
   return;
 }
 
+void
+TSRemapOSResponse(void *ih, TSHttpTxn rh, int os_response_type)
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: This function is practically identical to ``TSRemapdoRemap()``. Factor into a common subroutine.

File: plugins/experimental/ts_lua/ts_lua.c [hunk 4]
```diff
@@ -116,6 +116,69 @@ TSRemapDeleteInstance(void *ih)
...
+  ts_lua_set_cont_info(L, NULL);
+  if (lua_pcall(L, 0, 1, 0) != 0) {
+    TSError("[ts_lua] lua_pcall failed: %s", lua_tostring(L, -1));
+  }
+
+  lua_pop(L, 1);
+
+  if (http_ctx->has_hook) {
+    TSDebug(TS_LUA_DEBUG_TAG, "[%s] has txn hook -> adding txn close hook handler to release resources", __FUNCTION__);
+    TSHttpTxnHookAdd(rh, TS_HTTP_TXN_CLOSE_HOOK, contp);
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: If you do both remap and os_response, would you end up with a double free?

**Comment by shukitchan** [comment]
@jpeach , I have done the cleanup according to the comments. please take a look. thanks.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/436/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/329/ for details.

File: plugins/experimental/ts_lua/ts_lua_http.c [hunk 5]
```diff
@@ -29,8 +29,36 @@ typedef enum {
...
+                                            "TS_LUA_SRVSTATE_BAD_INCOMING_RESPONSE",
+                                            "TS_LUA_SRVSTATE_CONNECTION_ALIVE",
+                                            "TS_LUA_SRVSTATE_CONNECTION_CLOSED",
+                                            "TS_LUA_SRVSTATE_CONNECTION_ERROR",
+                                            "TS_LUA_SRVSTATE_INACTIVE_TIMEOUT",
+                                            "TS_LUA_SRVSTATE_OPEN_RAW_ERROR",
+                                            "TS_LUA_SRVSTATE_PARSE_ERROR",
+                                            "TS_LUA_SRVSTATE_TRANSACTION_COMPLETE",
+                                            "TS_LUA_SRVSTATE_CONGEST_CONTROL_CONGESTED_ON_F",
+                                            "TS_LUA_SRVSTATE_CONGEST_CONTROL_CONGESTED_ON_M"};
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua_http.c]: Try putting a ',' after the last element to get clang-format to be sensible.

File: plugins/experimental/ts_lua/ts_lua.c [hunk 6]
```diff
@@ -140,13 +141,15 @@ TSRemapDoRemap(void *ih, TSHttpTxn rh, TSRemapRequestInfo *rri)
...
-  http_ctx->txnp                = rh;
-  http_ctx->client_request_bufp = rri->requestBufp;
-  http_ctx->client_request_hdrp = rri->requestHdrp;
-  http_ctx->client_request_url  = rri->requestUrl;
-  http_ctx->rri                 = rri;
-  http_ctx->remap               = 1;
-  http_ctx->has_hook            = 0;
+  http_ctx->txnp     = rh;
+  http_ctx->has_hook = 0;
+  http_ctx->remap    = remap;
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: It looks like you don't really need the ``remap`` variable in the context; is it enough to have a non-NULL rri?

File: plugins/experimental/ts_lua/ts_lua.c [hunk 7]
```diff
@@ -117,7 +117,7 @@ TSRemapDeleteInstance(void *ih)
 }
 
 TSRemapStatus
-TSRemapDoRemap(void *ih, TSHttpTxn rh, TSRemapRequestInfo *rri)
+TSRemapPluginInit(void *ih, TSHttpTxn rh, TSRemapRequestInfo *rri)
```
Comment (jpeach) [inline plugins/experimental/ts_lua/ts_lua.c]: Make this static, and consider a name without the TS prefix.

File: plugins/experimental/ts_lua/ts_lua.c (hunk 6 above)
Comment (shukitchan) [inline plugins/experimental/ts_lua/ts_lua.c]: ha. that's true. didn't realize that till taking a step back

**Comment by shukitchan** [comment]
@jpeach cleanup is done accordingly. pls review.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/446/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/447/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/340/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/341/ for details.

**Comment by jpeach** [comment]
Loods good 👍

**Pull request merged**